cdap-examples/ClicksAndViews/src/main/java/co/cask/cdap/examples/clicksandviews/ClicksAndViewsMapReduce.java

```diff
@@ -0,0 +1,135 @@
+import com.google.common.base.Preconditions;
+import org.apache.hadoop.io.LongWritable;
+import org.apache.hadoop.io.NullWritable;
+import org.apache.hadoop.io.Text;
+import org.apache.hadoop.mapreduce.Job;
+import org.apache.hadoop.mapreduce.Mapper;
+import org.apache.hadoop.mapreduce.Reducer;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * A MapReduce program that reads from a CLICKS stream and a VIEWS stream. It performs a join across these two streams,
+ * based upon the viewId of the records. It then keeps track of how many clicks a particular view resulted in.
+ */
+public class ClicksAndViewsMapReduce extends AbstractMapReduce {
+  static final String NAME = "ClicksAndViewsMapReduce";
+
+  private static final Joiner TAB_JOINER = Joiner.on("\t");
+
+  @Override
+  public void configure() {
+    setName(NAME);
+    setMapperResources(new Resources(1024));
+    setReducerResources(new Resources(1024));
+  }
+
+  @Override
+  public void beforeSubmit(MapReduceContext context) throws Exception {
+    context.addInput(Input.ofStream(ClicksAndViews.CLICKS));
+    context.addInput(Input.ofStream(ClicksAndViews.VIEWS));
+
+    PartitionedFileSet joinedPFS = context.getDataset(ClicksAndViews.JOINED);
+    PartitionKey outputPartitionKey =
+      PartitionedFileSetArguments.getOutputPartitionKey(context.getRuntimeArguments(), joinedPFS.getPartitioning());
+
+    if (outputPartitionKey == null) {
+      outputPartitionKey = PartitionKey.builder().addLongField("runtime", context.getLogicalStartTime()).build();
+    }
+
+    Map<String, String> outputArgs = new HashMap<>();
+    PartitionedFileSetArguments.setOutputPartitionKey(outputArgs, outputPartitionKey);
+    context.addOutput(Output.ofDataset(ClicksAndViews.JOINED, outputArgs));
+
+    Job job = context.getHadoopJob();
+    job.setMapperClass(ImpressionKeyingMapper.class);
+    job.setReducerClass(JoiningReducer.class);
+  }
+
+  /**
+   * A Mapper which tags all of the records with the source that it is coming from.
+   * It also keys all of the output records with the viewId, so that a single reducer gets all of the records
+   * for a particular viewId.
+   */
+  public static class ImpressionKeyingMapper extends Mapper<LongWritable, Text, LongWritable, Text>
+    implements ProgramLifecycle<MapReduceTaskContext<LongWritable, Text>> {
+
+    private String inputName;
+
+    @Override
+    public void initialize(MapReduceTaskContext<LongWritable, Text> context) throws Exception {
+      inputName = context.getInputName();
+      Preconditions.checkNotNull(inputName);
+      Preconditions.checkArgument(ClicksAndViews.CLICKS.equals(inputName) || ClicksAndViews.VIEWS.equals(inputName));
+    }
+
+    @Override
+    public void destroy() {
+    }
+
+    @Override
+    protected void map(LongWritable key, Text value, Context context) throws IOException, InterruptedException {
+      String[] split = value.toString().split("\t");
```

Comment:
a little misleading to call this split. (A split is a different thing in MapReduce). Perhaps call it fields, or fragments?